### DISCUSSION TITLE
AWS hosting SSG ecommerce site

### DISCUSSION THREAD

**Discussion by lucaswong98** [vercel/next.js#38229] [Help]
Hi there,

My team is trying to host a next.js ecommerce app using AWS services because we have the aws credits in our account so we are thinking to utilise it. We are shifting from a fullstack C# nopcommerce app to RESTful api and a js frontend framework. The apis are ady built in and ready to go, so we are trying to use nextjs as our frontend app that provides features like seo and image optimization.

So far we come across AWS amplify service as the most commonly used service to host nextjs outside of vercel. However, we are not planning to use server side rendering and we are not building any api files in the nextjs app, as we ady a fully working api mentioned above, so the idea is to use SSG + ISG + Client side rendering only to handle the data fetched from api. 

So far we come across AWS amplify service as the most commonly used service to host nextjs outside of vercel. 

Question:
1. Does a SSG site on aws amplify still has the feature of incremental static generation (ISG) ? Because its not practical to rebuild the whole site if we have a product detail changed.
2. Does ISG requires aws lambda to work?
3. Is there any alternative to AWS amplify? We have a working EC2 VM so if possible we are trying to host our nextjs in the VM instead of using amplify

**Comment by icyJoseph** [comment]
> 1. Does a SSG site on aws amplify still has the feature of incremental static generation (ISG) ? Because its not practical to rebuild the whole site if we have a product detail changed.

I haven't used AWS amplify for Next.js apps, but I found this post, https://aws.amazon.com/blogs/mobile/host-a-next-js-ssr-app-with-real-time-data-on-aws-amplify/, from AWS themselves, where `revalidate: 1` is used. So I'll guess that it is possible to do ISG.

Although, it is a bit strange that the post doesn't mention this explicitly. 

```ts
// They do this
export async function getStaticProps (req) {
  const { DataStore } = withSSRContext(req)
  const { params } = req
  const { id } = params
  const post = await DataStore.query(Post, id)

  return {
    props: {
      post: JSON.parse(JSON.stringify(post))
    },
    revalidate: 1
  }
}
```

### Update

I found this bit of documentation, https://docs.aws.amazon.com/amplify/latest/userguide/server-side-rendering-amplify.html#supported-unsupported-features, where ISR is listed as supported by Amplify. 

> - Server-side rendered pages (SSR)
> - Static pages
> - API routes
> - Dynamic routes
> - Catch all routes
> - SSG (Static generation)
> - Incremental Static Regeneration (ISR)
> - Internationalized (i18n) sub-path routing
> - Environment variables
> - Image optimization. The size of the image can't exceed 1 MB. The AVIF and WebP image formats are not supported.

So I guess that's the full confirmation.

> 2. Does ISG requires aws lambda to work?

I don't know. I think `api/routes` do need this.

I'll try to read a bit more onto this. 

> When you deploy a Next.js app, Amplify creates Lambda@Edge functions to customize the content that CloudFront delivers.

Mmmm not sure what to make of this.

> 3. Is there any alternative to AWS amplify? We have a working EC2 VM so if possible we are trying to host our nextjs in the VM instead of using amplify

As per docs, https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#self-hosting-isr:

> Incremental Static Regeneration (ISR) works on [self-hosted Next.js sites](https://nextjs.org/docs/deployment#self-hosting) out of the box when you use next start.

We use an EC2 when deploying our app. This would be nearly the same as cloning the repo, installing, building, and running `next start`.

**Reply by lucaswong98** [reply]
Thank you so much for the answer, really appreciate it.

One more question to the ISR part tho, for now, we know that AWS Amplify works with ISR out of the box, but can we use it on static sites or when using `next export` tho? Because the document that you posted is sharing about SSR deployment

**Reply by icyJoseph** [reply]
https://nextjs.org/docs/advanced-features/static-html-export#unsupported-features

`next export` doesn't support ISR, the export mode is just a static HTML page, there's no server.

**Reply by lucaswong98** [reply]
Oh right, forgot about that, `next export` is meant to generate HTML pages on build time
